File: decl-ts/src/qengine/rounds.ts
// Reference-round reuse over the value layer. The dependency graph includes
// member reads (also cache hits), roots, frozen member reads and edge answers.
// A failed eligibility check keeps the fresh-round evaluator as the fallback.
import type { Engine } from '../engine.ts';
import { isArr, isMap, isRec, isRef, isQ, isRange, pathStr, valueEq } from '../semantics.ts';
import type { Env, RecInst, Seg, Value } from '../semantics.ts';

type Edge = Map<string, { path: Seg[]; refs: Seg[][] }>;
class IneligibleSnapshot extends Error {}

const slotKey = (inst: RecInst, name: string) => `${pathStr(inst.path)}.${name}`;

/** Compare the observed container shape; member values have their own reads.
 * Context-dependent results are handled separately by snapshotResult. */
function same(a: Value, b: Value): boolean {
  if (a === b) return true;
  if (isRec(a) && isRec(b)) {
    const as = [...a.slots],
      bs = [...b.slots];
    return (
      a.typeName === b.typeName &&
      a.rt === b.rt &&
      pathStr(a.path) === pathStr(b.path) &&
      as.length === bs.length &&
      as.every(
        ([n, x], i) =>
          n === bs[i][0] && x.kind === bs[i][1].kind && !!x.hidden === !!bs[i][1].hidden,
      ) &&
      a.entryOrder.length === b.entryOrder.length &&
      a.entryOrder.every((n: string, i: number) => n === b.entryOrder[i]) &&
      same({ __map: true, entries: a.extras }, { __map: true, entries: b.extras })
    );
  }
  if (isArr(a) && isArr(b))
    return (
      pathStr(a.path) === pathStr(b.path) &&
      a.items.length === b.items.length &&
      a.items.every((v: Value, i: number) => same(v, b.items[i]))
    );
  if (isMap(a) && isMap(b)) {
    const ae = [...a.entries],
      be = [...b.entries];
    return (
      pathStr(a.path ?? []) === pathStr(b.path ?? []) &&
      ae.length === be.length &&
      ae.every(([k, v], i) => k === be[i][0] && same(v, be[i][1]))
    );
  }
  if (isRec(a) || isRec(b) || isArr(a) || isArr(b) || isMap(a) || isMap(b)) return false;
  return valueEq(a, b);
}

/** A cached scalar can survive an epoch change. A forwarded inverse reference
 * or frozen record view cannot, even if its canonical path is unchanged. */
function snapshotResult(v: Value, eng: Engine, prefix: string, seen = new Set<object>()): boolean {
  if (!v || typeof v !== 'object' || seen.has(v)) return false;
  seen.add(v);
  if (isRef(v)) return !!(v.inverseRef || v.snap);
  if (isArr(v)) return v.items.some((x: Value) => snapshotResult(x, eng, prefix, seen));
  if (isMap(v)) return [...v.entries.values()].some((x) => snapshotResult(x, eng, prefix, seen));
  if (isRec(v)) {
    const path = pathStr(v.path);
    if (
      ((v as any).eng && (v as any).eng !== eng) ||
      !(path === prefix || path.startsWith(prefix + '.') || path.startsWith(prefix + '['))
    )
      return true;
    return (
      [...v.slots.values()].some((s) => snapshotResult(s.value, eng, prefix, seen)) ||
      [...v.extras.values()].some((x) => snapshotResult(x, eng, prefix, seen))
    );
  }
  return false;
}

/** Statistics describe work, without adding user-visible diagnostics. */
export class RoundCache {
  /** Skip dependency recording when the loaded module graph has no reference
   * queries. A missed optimization still follows the complete fresh driver. */
  static needed(env: Env, seen = new Set<Env>()): boolean {
    if (seen.has(env)) return false;
    seen.add(env);
    const contains = (node: unknown): boolean =>
      node !== null &&
      typeof node === 'object' &&
      (('e' in node && node.e === 'referrers') || Object.values(node).some(contains));
    const asts = [
      ...env.typeAsts.values(),
      ...env.funcs.values(),
      ...env.inputs.values(),
      ...env.outputs,
      ...env.unitDecls.values(),
      ...[...env.consts.values()].map((c) => [c.expr, c.type]),
    ];
    return (
      asts.some(contains) ||
      [...env.imports.values()].some((im) => this.needed(im.env, seen)) ||
      [...env.namespaces.values()].some((ns) => this.needed(ns.env, seen))
    );
  }

  rounds = 0;
  reusedRounds = 0;
  retainedRecords = 0;
  invalidatedSlots = 0;
  readonly cleanRecords = new WeakSet<RecInst>();
  private rebased = new WeakMap<object, Value>();

  /** Reused referrer values denote the new round's frozen universe. Frozen
   * copies have roundRef cleared, so their references keep their old owner. */
  value(v: Value, previous: Engine | null): Value {
    if (!previous || !v || typeof v !== 'object' || isRec(v)) return v;
    const cached = this.rebased.get(v);
    if (cached !== undefined) return cached;
    let result = v;
    if (isRef(v) && v.roundRef) result = { ...v, snap: previous };
    else if (isArr(v)) {
      const items = v.items.map((x: Value) => this.value(x, previous));
      if (items.some((x: Value, i: number) => x !== v.items[i])) result = { ...v, items };
    } else if (isMap(v)) {
      const entries = new Map<string, Value>();
      let changed = false;
      for (const [k, x] of v.entries) {
        const y = this.value(x, previous);
        entries.set(k, y);
        changed ||= y !== x;
      }
      if (changed) result = { ...v, entries };
    }
    this.rebased.set(v, result);
    return result;
  }

  advance(eng: Engine, edges: Map<string, Edge>): Engine | null {
    this.rounds++;
    const env = eng.env;
    // Unfinished values and diagnostics require the fresh evaluator's exact
    // retry/reporting behavior. Never guess that an incomplete value is pure.
    if (
      env.diagnostics.length ||
      env.registry.some((r) =>
        [...r.slots.values()].some((s) => s.state !== 'ok' && s.state !== 'absent'),
      )
    )
      return null;
    const priorRecords = new Map((eng.prev?.frozenRegistry ?? []).map((r) => [pathStr(r.path), r]));
    const records = new Map(env.registry.map((r) => [pathStr(r.path), r]));
    if (records.size !== env.registry.length) return null;
    const reverse = new Map<string, Set<string>>();
    const pending: string[] = [];
    const indexes = new Map<Edge, Map<string, string[]>>();
    const answer = (edge: Edge | undefined, target: string): string[] => {
      if (!edge) return [];
      let index = indexes.get(edge);
      if (!index) {
        index = new Map();
        indexes.set(edge, index);
        for (const [source, row] of edge)
          for (const target of new Set(row.refs.map((p) => pathStr(p))))
            (index.get(target) ?? index.set(target, []).get(target)!).push(source);
        for (const paths of index.values()) paths.sort();
      }
      return index.get(target) ?? [];
    };
    for (const [reader, deps] of eng.reads)
      for (const dep of deps) {
        (reverse.get(dep) ?? reverse.set(dep, new Set()).get(dep)!).add(reader);
        if (dep.startsWith('edge:')) {
          // Referrer answers include an empty result: the first incoming edge
          // invalidates the readers that previously found no candidate at all.
          const parts = dep.slice(5).split('|');
          const key = `${parts[0]}|${parts[1]}`,
            target = parts.slice(2).join('|');
          const a = answer(eng.snap?.edges.get(key), target),
            b = answer(edges.get(key), target);
          if (a.length !== b.length || a.some((p, i) => p !== b[i])) pending.push(dep);
        } else if (dep === 'round:nested') {
          pending.push(dep);
        } else if (dep === 'round:reference') {
          const slot = eng.slotsByKey.get(reader);
          const prefix = reader.startsWith('root:') ? reader.slice(5) : reader;
          const value = slot ? slot.inst.slots.get(slot.name)?.value : env.roots.get(prefix);
          if (reader.startsWith('const:') || snapshotResult(value, eng, prefix))
            pending.push(reader);
        } else if (dep.startsWith('snapshot:')) {
          const current = eng.slotsByKey.get(dep.slice(9));
          const a =
            current && priorRecords.get(pathStr(current.inst.path))?.slots.get(current.name);
          const b = current?.inst.slots.get(current.name);
          const av = eng.prev?.snapshotValue ? eng.prev.snapshotValue(a?.value) : a?.value;
          if (!a || !b || a.state !== b.state || !same(av, b.value)) pending.push(dep);
        }
      }
    // Constants may own typed records and lexical closures; conservatively
    // re-open their readers until constant inputs have independent ownership.

    const invalid = new Set<string>();
    const dropped = new Set<RecInst>();
    const roots = new Set<string>();
    const queue = pending;
    const subtrees = new Map<string, RecInst[]>();
    for (const inst of records.values())
      for (let i = 1; i <= inst.path.length; i++) {
        const prefix = pathStr(inst.path.slice(0, i));
        (subtrees.get(prefix) ?? subtrees.set(prefix, []).get(prefix)!).push(inst);
      }
    const addSubtree = (path: Seg[]) => {
      const prefix = pathStr(path);
      for (const inst of subtrees.get(prefix) ?? []) {
        if (dropped.has(inst)) continue;
        dropped.add(inst);
        for (const name of inst.slots.keys()) queue.push(slotKey(inst, name));
      }
    };
    while (queue.length) {
      const key = queue.pop()!;
      if (invalid.has(key)) continue;
      invalid.add(key);
      if (key.startsWith('const:')) return null;
      for (const reader of reverse.get(key) ?? []) queue.push(reader);
      if (key.startsWith('root:')) {
        const name = key.slice(5);
        roots.add(name);
        addSubtree([name]);
      } else {
        const slot = eng.slotsByKey.get(key);
        if (slot) addSubtree([...slot.inst.path, slot.name]);
      }
    }
    // Rebinding an earlier root could change the forcing order of retained
    // later roots. Keep the fresh driver for that case.
    let rebound = false;
    for (const name of env.roots.keys()) {
      rebound ||= roots.has(name);
      if (rebound && !roots.has(name)) return null;
    }
    const dirty = new Set<RecInst>();
    for (const key of invalid) {
      const found = eng.slotsByKey.get(key);
      for (let inst = found?.inst ?? null; inst; inst = inst.parent) dirty.add(inst);
    }
    let snapshot: Engine;
    try {
      snapshot = this.freeze(eng);
    } catch (error) {
      if (error instanceof IneligibleSnapshot) return null;
      throw error;
    }
    for (const inst of dropped)
      for (const name of inst.slots.keys()) {
        const key = slotKey(inst, name);
        eng.reads.delete(key);
        eng.slotsByKey.delete(key);
      }
    const copiedSlots = new Set<RecInst>();
    for (const key of invalid) {
      const found = eng.slotsByKey.get(key);
      if (found) {
        if (!copiedSlots.has(found.inst)) {
          found.inst.slots = new Map(found.inst.slots);
          copiedSlots.add(found.inst);
        }
        const slot = found.inst.slots.get(found.name)!;
        if (slot.compute) {
          found.inst.slots.set(found.name, { ...slot, state: 'unforced', value: undefined });
          this.invalidatedSlots++;
        }
      }
      eng.reads.delete(key);
    }
    env.registry.splice(0, env.registry.length, ...env.registry.filter((r) => !dropped.has(r)));
    for (const inst of env.registry) {
      if (dirty.has(inst)) this.cleanRecords.delete(inst);
      else this.cleanRecords.add(inst);
    }
    for (const name of roots) env.roots.delete(name);
    eng.deferredSlots = [];
    eng.deferredRoots = [];
    eng.roundRoots = roots;
    eng.phase = 1;
    eng.snap = null;
    eng.queried.clear();
    for (const deps of eng.reads.values())
      for (const dep of deps)
        if (dep.startsWith('edge:')) eng.queried.add(dep.slice(5).split('|').slice(0, 2).join('|'));
    eng.refIndex.clear();
    eng.computingEdges.clear();
    eng.edgeBases = [];
    for (const reset of eng.roundResets) reset();

    this.rebased = new WeakMap();
    for (const [name, value] of env.roots) env.roots.set(name, this.value(value, snapshot));
    this.retainedRecords += env.registry.length;
    this.reusedRounds++;
    return snapshot;
  }

  private freeze(eng: Engine): Engine {
    const frozen = Object.assign(Object.create(Object.getPrototypeOf(eng)), eng) as Engine;
    const copies = new WeakMap<object, Value>();
    const previous = eng.prev;
    const normalizer = new RoundCache();
    const copy = (raw: Value): Value => {
      const v = normalizer.value(raw, previous);
      if (!v || typeof v !== 'object') return v;
      const hit = copies.get(v);
      if (hit) return hit;
      if (isRef(v)) return v.roundRef ? { ...v, roundRef: false } : v;
      if (isQ(v) || isRange(v)) return { ...v };
      if (v.__jobj) return v; // lexical JSON is immutable input data
      if (isRec(v)) {
        if ((v as any).eng && (v as any).eng !== eng) return v;
        const r = { ...v, eng: frozen };
        copies.set(v, r);
        r.parent = v.parent ? copy(v.parent) : null;
        return r;
      }
      if (isArr(v)) {
        const a = { ...v, items: [] as Value[] };
        copies.set(v, a);
        a.items = v.items.map(copy);
        if (a.items.every((x: Value, i: number) => x === v.items[i])) {
          copies.set(v, v);
          return v;
        }
        return a;
      }
      if (isMap(v)) {
        const m = { ...v, entries: new Map<string, Value>() };
        copies.set(v, m);
        let changed = false;
        for (const [k, x] of v.entries) {
          const y = copy(x);
          m.entries.set(k, y);
          changed ||= x !== y;
        }
        if (!changed) {
          copies.set(v, v);
          return v;
        }
        return m;
      }
      // Lazy prevalues and closures capture execution context; retain the
      // fresh-round behavior until they can be represented as snapshot data.
      throw new IneligibleSnapshot('context-bearing snapshot');
    };
    frozen.frozenRegistry = eng.env.registry.map(copy);
    frozen.frozenRoots = new Map([...eng.env.roots].map(([k, v]) => [k, copy(v)]));
    // Validate and memoize container copies once. Record slot maps are shared
    // with the snapshot until the live owner replaces a slot (copy-on-write).
    for (const inst of eng.env.registry) {
      for (const slot of inst.slots.values()) copy(slot.value);
      for (const value of inst.extras.values()) copy(value);
    }
    frozen.slotsByKey = new Map();
    frozen.snapshotValue = copy;
    frozen.roundCache = null;
    frozen.roundRoots = null;
    frozen.track = false;
    frozen.computing = [];
    frozen.reads = new Map();
    return frozen;
  }
}

File: decl-py/src/decl/qengine/rounds.py
"""Reference-round reuse over the value layer; mirrors qengine/rounds.ts.

Member reads, roots, frozen member reads and edge answers are dependencies.
Context-bearing or unfinished snapshots retain fresh-round evaluation.
"""

from __future__ import annotations

from copy import copy
from typing import TYPE_CHECKING, Any

from decl.semantics import ArrV, JObj, MapV, Quantity, RangeV, RecInst, Ref, path_str, value_eq

if TYPE_CHECKING:
    from decl.engine import Engine


def slot_key(inst: RecInst, name: str) -> str:
    return f"{path_str(inst.path)}.{name}"


def same(a: Any, b: Any) -> bool:
    """Container shape equality; member values have separate read dependencies."""
    if a is b:
        return True
    if isinstance(a, RecInst) and isinstance(b, RecInst):
        return (
            a.type_name == b.type_name
            and a.rt is b.rt
            and a.path == b.path
            and [(n, s.kind, s.hidden) for n, s in a.slots.items()]
            == [(n, s.kind, s.hidden) for n, s in b.slots.items()]
            and a.entry_order == b.entry_order
            and list(a.extras) == list(b.extras)
            and all(same(v, b.extras[k]) for k, v in a.extras.items())
        )
    if isinstance(a, ArrV) and isinstance(b, ArrV):
        return (
            a.path == b.path
            and len(a.items) == len(b.items)
            and all(same(x, y) for x, y in zip(a.items, b.items, strict=True))
        )
    if isinstance(a, MapV) and isinstance(b, MapV):
        return (
            a.path == b.path
            and list(a.entries) == list(b.entries)
            and all(same(v, b.entries[k]) for k, v in a.entries.items())
        )
    if isinstance(a, (RecInst, ArrV, MapV)) or isinstance(b, (RecInst, ArrV, MapV)):
        return False
    return value_eq(a, b)


def snapshot_result(v: Any, eng: Engine, prefix: str, seen: set[int] | None = None) -> bool:
    """Forwarded inverse references and frozen record views depend on the epoch."""
    if not isinstance(v, (Ref, RecInst, ArrV, MapV)):
        return False
    seen = set() if seen is None else seen
    if id(v) in seen:
        return False
    seen.add(id(v))
    if isinstance(v, Ref):
        return bool(v.inverse or v.snap)
    if isinstance(v, ArrV):
        return any(snapshot_result(x, eng, prefix, seen) for x in v.items)
    if isinstance(v, MapV):
        return any(snapshot_result(x, eng, prefix, seen) for x in v.entries.values())
    path = path_str(v.path)
    if (v.eng is not None and v.eng is not eng) or not (
        path == prefix or path.startswith(prefix + ".") or path.startswith(prefix + "[")
    ):
        return True
    return any(snapshot_result(s.value, eng, prefix, seen) for s in v.slots.values()) or any(
        snapshot_result(x, eng, prefix, seen) for x in v.extras.values()
    )


class IneligibleSnapshot(Exception):
    """The snapshot contains a value requiring fresh-round evaluation."""


class RoundCache:
    """A retained dependency graph uses the slots as its single value cache."""

    @staticmethod
    def needed(env: Any, seen: set[Any] | None = None) -> bool:
        from decl.semantics import mentions_referrers

        seen = set() if seen is None else seen
        if env in seen:
            return False
        seen.add(env)
        asts = [
            *env.type_asts.values(),
            *env.funcs.values(),
            *env.inputs.values(),
            *env.outputs,
            *env.unit_decls.values(),
            *[[c["expr"], c.get("type")] for c in env.consts.values()],
        ]
        return (
            mentions_referrers(asts)
            or any(RoundCache.needed(im["env"], seen) for im in env.imports.values())
            or any(RoundCache.needed(ns["env"], seen) for ns in env.namespaces.values())
        )

    def __init__(self) -> None:
        self.clean_records: set[RecInst] = set()
        self.rounds = 0
        self.reused_rounds = 0
        self.retained_records = 0
        self.invalidated_slots = 0
        self.rebased: dict[int, tuple[Any, Any]] = {}

    def value(self, v: Any, previous: Engine | None) -> Any:
        if previous is None or not isinstance(v, (Ref, ArrV, MapV)):
            return v
        hit = self.rebased.get(id(v))
        if hit is not None:
            return hit[1]
        result = v
        if isinstance(v, Ref) and v.round_ref:
            result = Ref(v.segs, previous, True)
        elif isinstance(v, ArrV):
            items = [self.value(x, previous) for x in v.items]
            if any(x is not y for x, y in zip(items, v.items, strict=True)):
                result = ArrV(items, v.path)
        elif isinstance(v, MapV):
            entries = {k: self.value(x, previous) for k, x in v.entries.items()}
            if any(x is not v.entries[k] for k, x in entries.items()):
                result = MapV(entries, v.path)
        self.rebased[id(v)] = (v, result)
        return result

    def advance(self, eng: Engine, edges: dict[str, Any]) -> Engine | None:
        self.rounds += 1
        env = eng.env
        if env.diagnostics or any(
            s.state not in ("ok", "absent") for r in env.registry for s in r.slots.values()
        ):
            return None
        prior_records = (
            {path_str(r.path): r for r in (eng.prev.frozen_registry or [])} if eng.prev else {}
        )
        records = {path_str(r.path): r for r in env.registry}
        if len(records) != len(env.registry):
            return None
        reverse: dict[str, set[str]] = {}
        pending: list[str] = []
        indexes: dict[int, dict[str, list[str]]] = {}

        def answer(edge: dict[str, Any], target: str) -> list[str]:
            index = indexes.get(id(edge))
            if index is None:
                index = {}
                indexes[id(edge)] = index
                for source, row in edge.items():
                    for dest in {path_str(p) for p in row["refs"]}:
                        index.setdefault(dest, []).append(source)
                for paths in index.values():
                    paths.sort()
            return index.get(target, [])

        empty: dict[str, Any] = {}
        for reader, deps in eng.reads.items():
            for dep in deps:
                reverse.setdefault(dep, set()).add(reader)
                if dep.startswith("edge:"):
                    type_name, member, target = dep[5:].split("|", 2)
                    key = f"{type_name}|{member}"
                    prior = eng.snap["edges"].get(key, empty) if eng.snap else empty
                    if answer(prior, target) != answer(edges.get(key, empty), target):
                        pending.append(dep)
                elif dep == "round:nested":
                    pending.append(dep)
                elif dep == "round:reference":
                    slot = eng.slots_by_key.get(reader)
                    prefix = reader[5:] if reader.startswith("root:") else reader
                    value = slot[0].slots[slot[1]].value if slot else env.roots.get(prefix)
                    if reader.startswith("const:") or snapshot_result(value, eng, prefix):
                        pending.append(reader)
                elif dep.startswith("snapshot:"):
                    cur = eng.slots_by_key.get(dep[9:])
                    old = prior_records.get(path_str(cur[0].path)) if cur else None
                    a = old.slots.get(cur[1]) if old and cur else None
                    b = cur[0].slots.get(cur[1]) if cur else None
                    av = a.value if a else None
                    if eng.prev and eng.prev.snapshot_value:
                        av = eng.prev.snapshot_value(av)
                    if a is None or b is None or a.state != b.state or not same(av, b.value):
                        pending.append(dep)
        invalid: set[str] = set()
        dropped: set[RecInst] = set()
        roots: set[str] = set()
        subtrees: dict[str, list[RecInst]] = {}
        for inst in records.values():
            for i in range(1, len(inst.path) + 1):
                subtrees.setdefault(path_str(inst.path[:i]), []).append(inst)

        def add_subtree(path: list[Any]) -> None:
            for inst in subtrees.get(path_str(path), []):
                if inst in dropped:
                    continue
                dropped.add(inst)
                pending.extend(slot_key(inst, name) for name in inst.slots)

        while pending:
            key = pending.pop()
            if key in invalid:
                continue
            invalid.add(key)
            if key.startswith("const:"):
                return None
            pending.extend(reverse.get(key, ()))
            if key.startswith("root:"):
                name = key[5:]
                roots.add(name)
                add_subtree([name])
            else:
                found = eng.slots_by_key.get(key)
                if found:
                    add_subtree([*found[0].path, found[1]])
        rebound = False
        for name in env.roots:
            rebound = rebound or name in roots
            if rebound and name not in roots:
                return None
        dirty: set[RecInst] = set()
        for key in invalid:
            found = eng.slots_by_key.get(key)
            inst = found[0] if found else None
            while inst is not None:
                dirty.add(inst)
                inst = inst.parent
        try:
            snapshot = self.freeze(eng)
        except IneligibleSnapshot:
            return None
        for inst in dropped:
            for name in inst.slots:
                key = slot_key(inst, name)
                eng.reads.pop(key, None)
                eng.slots_by_key.pop(key, None)
        copied_slots: set[RecInst] = set()
        for key in invalid:
            found = eng.slots_by_key.get(key)
            if found:
                if found[0] not in copied_slots:
                    found[0].slots = dict(found[0].slots)
                    copied_slots.add(found[0])
                inst, name = found
                slot = inst.slots[name]
                if slot.compute is not None:
                    fresh = copy(slot)
                    fresh.state, fresh.value = "unforced", None
                    inst.slots[name] = fresh
                    self.invalidated_slots += 1
            eng.reads.pop(key, None)
        env.registry[:] = [r for r in env.registry if r not in dropped]
        self.clean_records = set(env.registry) - dirty
        for name in roots:
            env.roots.pop(name, None)
        eng.deferred_slots = []
        eng.deferred_roots = []
        eng.round_roots = roots
        eng.phase = 1
        eng.snap = None
        eng.queried = {
            "|".join(dep[5:].split("|", 2)[:2])
            for deps in eng.reads.values()
            for dep in deps
            if dep.startswith("edge:")
        }
        eng.ref_index.clear()
        eng.computing_edges.clear()
        eng.edge_bases = []
        for reset in eng.round_resets:
            reset()
        self.rebased = {}
        for name, value in env.roots.items():
            env.roots[name] = self.value(value, snapshot)
        self.retained_records += len(env.registry)
        self.reused_rounds += 1
        return snapshot

    def freeze(self, eng: Engine) -> Engine:
        frozen = copy(eng)
        copies: dict[int, Any] = {}
        previous = eng.prev
        normalizer = RoundCache()

        def clone(raw: Any) -> Any:
            v = normalizer.value(raw, previous)
            if type(v) in (type(None), bool, int, float, str):
                return v
            if id(v) in copies:
                return copies[id(v)]
            if isinstance(v, Ref):
                return Ref(v.segs, v.snap, inverse=True) if v.round_ref else v
            if isinstance(v, (Quantity, RangeV, JObj)):
                return v
            if isinstance(v, RecInst):
                if v.eng is not None and v.eng is not eng:
                    return v
                r = RecInst(v.type_name, v.rt, v.path, None)
                copies[id(v)] = r
                r.eng, r.menv = frozen, v.menv
                r.entry_order = v.entry_order
                r.parent = clone(v.parent) if v.parent else None
                r.slots = v.slots
                r.extras = v.extras
                return r
            if isinstance(v, ArrV):
                a = ArrV([], v.path)
                copies[id(v)] = a
                a.items = [clone(x) for x in v.items]
                if all(x is y for x, y in zip(a.items, v.items, strict=True)):
                    copies[id(v)] = v
                    return v
                return a
            if isinstance(v, MapV):
                m = MapV({}, v.path)
                copies[id(v)] = m
                m.entries = {k: clone(x) for k, x in v.entries.items()}
                if all(x is v.entries[k] for k, x in m.entries.items()):
                    copies[id(v)] = v
                    return v
                return m
            from decl.semantics import ABSENT

            if v is ABSENT:
                return v
            raise IneligibleSnapshot("context-bearing snapshot")

        frozen.frozen_registry = [clone(r) for r in eng.env.registry]
        frozen.frozen_roots = {k: clone(v) for k, v in eng.env.roots.items()}
        for inst in eng.env.registry:
            for slot in inst.slots.values():
                clone(slot.value)
            for value in inst.extras.values():
                clone(value)
        frozen.slots_by_key = {}
        frozen.snapshot_value = clone
        frozen.round_cache = None
        frozen.round_roots = None
        frozen.track = False
        frozen.computing = []
        frozen.reads = {}
        return frozen
